### PR TITLE
Enhance local experiments

### PR DESCRIPTION
**Pull request by andreafioraldi**
Followup to my `--concurrent-builds` PR to fix the local experiments.
Here there are several contributions:
+ Improve syscalls speed in the containers with `--security-opt seccomp=unconfined` (see the AFL++ system config script) when running runners containers
+ Pin runners and measures to specific cores with `--cpuset-cpus` for runners and psutil for measures.
+ Allow the user to specify how many cores the measures must use with `--measurers-cpus` in run_experiment.py
+ Allow the user to specify how many cores the runners must use with `--runners-cpus` in run_experiment.py

Now, for instance, I can easily run a local experiment even on my desktop without any explosion of docker or OOM with

```
PYTHONPATH=. python3 experiment/run_experiment.py \
--experiment-config foo.yaml \
--benchmarks libpng-1.2.56 jsoncpp_jsoncpp_fuzzer \
--experiment-name foo \
--concurrent-builds 2 \
--measurers-cpus 4 \
--runners-cpus 4 \
--fuzzers afl aflfast \
```
The outcome is

Screenshot at 2022-03-17 16-00-40

Before was a mess with cores not used at 100% and overlapped with measurers.

I tested it with and without the new arguments and it works, but before merging you should try to start a remote experiment even if in theory I didn't touch the remote part.

### CODE REVIEW

**Comment by andreafioraldi** [comment]
I updated the doc too.

File: experiment/resources/runner-startup-script-template.sh [hunk 1]
```diff
@@ -52,4 +53,5 @@ docker run \
 -e LOCAL_EXPERIMENT={{local_experiment}} \
 {% if not local_experiment %}--name=runner-container {% endif %}\
 --cap-add SYS_NICE --cap-add SYS_PTRACE \
+--security-opt seccomp=unconfined \
```
Comment (jonathanmetzman) [inline experiment/resources/runner-startup-script-template.sh:56]: So what would be the downsides of this?

File: experiment/run_experiment.py [hunk 2]
```diff
@@ -235,9 +237,13 @@ def start_experiment(  # pylint: disable=too-many-arguments
     config['oss_fuzz_corpus'] = oss_fuzz_corpus
     config['description'] = description
     config['concurrent_builds'] = concurrent_builds
+    config['measurers_cpus'] = measurers_cpus
+    config['runners_cpus'] = runners_cpus
     config['runner_machine_type'] = config.get('runner_machine_type',
                                                'n1-standard-1')
     config['runner_num_cpu_cores'] = config.get('runner_num_cpu_cores', 1)
+    assert (runners_cpus is None or
+            runners_cpus >= config['runner_num_cpu_cores'])
```
Comment (jonathanmetzman) [inline experiment/run_experiment.py:246]: Why enforce this?

File: experiment/run_experiment.py [hunk 3]
```diff
@@ -549,8 +563,27 @@ def main():
...
+            parser.error('The runners cpus argument must be a positive number')
+        runners_cpus = int(runners_cpus)
+    measurers_cpus = args.measurers_cpus
+    if measurers_cpus is not None:
+        if not measurers_cpus.isdigit():
+            parser.error(
+                'The measurers cpus argument must be a positive number')
+        if runners_cpus is None:
+            parser.error(
+                'With the measurers cpus argument you need to specity the'
```
Comment (jonathanmetzman) [inline experiment/run_experiment.py:580]: nit: `specify`

**Review by jonathanmetzman — COMMENTED** [review]
This is great stuff. I took a first pass.
I'll try to finish reviewing tomorrow once I understand the cpuset stuff better.

File: experiment/resources/runner-startup-script-template.sh (hunk 1 above)
Comment (andreafioraldi) [reply experiment/resources/runner-startup-script-template.sh:56]: https://stackoverflow.com/questions/60840320/docker-50-performance-hit-on-cpu-intensive-code/61053027#61053027

No downsides imo, it's not a service exposed running on docker that has security boundaries

File: experiment/run_experiment.py (hunk 2 above)
Comment (andreafioraldi) [reply experiment/run_experiment.py:246]: well, if i say that runners must run on 2 cores but then each runner needs 3 cores it's not ok.

**Review by jonathanmetzman — APPROVED** [review]
lgtm

**Review by jonathanmetzman — CHANGES_REQUESTED** [review]

File: experiment/dispatcher.py [hunk 4]
```diff
@@ -101,6 +101,8 @@ def __init__(self, experiment_config_filepath: str):
         self.experiment_name = self.config['experiment']
         self.git_hash = self.config['git_hash']
         self.concurrent_builds = self.config['concurrent_builds']
+        self.measurers_cpus = self.config['measurers_cpus']
```
[the remark this answers is not in the document]
Comment (andreafioraldi) [reply experiment/dispatcher.py]: ok, I removed concurrent_builds too

**Review by jonathanmetzman — APPROVED** [review]
lgtm

**Comment by jiradeto** [comment]
Just come across and found this PR is very helpful as I often experienced this issue for the local experiment. Thanks a lot for doing this @andreafioraldi. 

I am wondering what would happen if there are multiple schedulers pinned to the same cpu core. For instance, if 4 schedulers with cpuset-cpus = 0, would cpu resource be evenly allocated 25% for each scheduler of CPU 0?

**Comment by apach301** [comment]
Thanks @andreafioraldi for this PR, looks like it should really help to scale experiments on different machines. But when I tried these options on local experiments it worked not as I expected. I launched 5 benchmarks with 2 fuzzers and 5 trials, so I scheduled 50 cpus on measurers and runners accordingly:
```
DOCKER_BUILDKIT=1 PYTHONPATH=. python3 experiment/run_experiment.py -a \
--experiment-config experiment-libfuzzer.yaml \
--benchmarks lcms-2017-03-21 openssl_x509 openthread-2019-12-23\
                       vorbis-2017-12-11 woff2-2016-05-06 \
--experiment-name try-new-options-1 \
--fuzzers libfuzzer sydr_libfuzzer \
--concurrent-builds 2 \
--measurers-cpus 50 \
--runners-cpus 50
```
But after building all dockers only 11 cores were used at 100%, while other not used at all:

fuzzbench_htop_1

Sometimes cores from 50 to 100 starts working too, I guess that's because measurers have been activated. Doesn't all 50 cores for runners should be busy (all benchmarks and trials are working at that time)? Logs says that all requested cores have been scheduled for runners and measurers:
```
INFO:root:Done building fuzzer benchmarks.
INFO:root:Starting scheduler.
INFO:root:Scheduling runners from core 0 to 49.
INFO:root:Finding trials to schedule.
INFO:root:Starting trials.
INFO:root:Start trial 4123.
...
INFO:root:Start trial 4154.
INFO:root:Start measuring.
INFO:root:Start measure_loop.
INFO:root:Scheduling measurers from core 50 to 99.
INFO:root:Start trial 4143.
...
INFO:root:Start trial 4169.
INFO:root:Done starting trials.
INFO:root:Measuring all trials.
INFO:root:Measuring all trials.
INFO:root:Measuring all trials.
```

**Comment by andreafioraldi** [comment]
> Thanks @andreafioraldi for this PR, looks like it should really help to scale experiments on different machines. But when I tried these options on local experiments it worked not as I expected. I launched 5 benchmarks with 2 fuzzers and 5 trials, so I scheduled 50 cpus on measurers and runners accordingly:
> 
> ```
> DOCKER_BUILDKIT=1 PYTHONPATH=. python3 experiment/run_experiment.py -a \
> --experiment-config experiment-libfuzzer.yaml \
> --benchmarks lcms-2017-03-21 openssl_x509 openthread-2019-12-23\
>                        vorbis-2017-12-11 woff2-2016-05-06 \
> --experiment-name try-new-options-1 \
> --fuzzers libfuzzer sydr_libfuzzer \
> --concurrent-builds 2 \
> --measurers-cpus 50 \
> --runners-cpus 50
> ```
> 
> But after building all dockers only 11 cores were used at 100%, while other not used at all:
> 
> fuzzbench_htop_1
> 
> Sometimes cores from 50 to 100 starts working too, I guess that's because measurers have been activated. Doesn't all 50 cores for runners should be busy (all benchmarks and trials are working at that time)? Logs says that all requested cores have been scheduled for runners and measurers:
> 
> ```
> INFO:root:Done building fuzzer benchmarks.
> INFO:root:Starting scheduler.
> INFO:root:Scheduling runners from core 0 to 49.
> INFO:root:Finding trials to schedule.
> INFO:root:Starting trials.
> INFO:root:Start trial 4123.
> ...
> INFO:root:Start trial 4154.
> INFO:root:Start measuring.
> INFO:root:Start measure_loop.
> INFO:root:Scheduling measurers from core 50 to 99.
> INFO:root:Start trial 4143.
> ...
> INFO:root:Start trial 4169.
> INFO:root:Done starting trials.
> INFO:root:Measuring all trials.
> INFO:root:Measuring all trials.
> INFO:root:Measuring all trials.
> ```

Mmmmm dunno never tried with this amount of cores. I cannot look into it atm, try to debug it a bit yourself (simply see which function I patched in this PR).

**Comment by apach301** [comment]
Ok, I'll try to look into the code more deeply. At first glance it should work correctly, but I think the problem may be in the affinity. In my case all runners have started but apparently they were mapped on 12 cores.

**Comment by apach301** [comment]
Added some debug prints. Looks like CPUSET is set after _start_trial, so several trials are runned on same core:

cpuset_failure

**Comment by apach301** [comment]
We determined the problem source. Python starts a thread pool. Each worker accept task to call Popen runner and exit without waiting for it. So, starmap schedules trials for a subset of cores. All work is immediately done because no one is waiting for the child process in local experiment. So, multiple trials run on one core with overcommit. We are trying to figure out the patch. However, the main problem is that we need to both wait for child processes and update trial start time.

**Comment by SweetVishnya** [comment]
We still do not get why we need a thread pool for local experiment. Each worker exits immediately. We could just Popen N trials in a loop. Then store processes in some list and wait for all of them to finish. Then start another batch of N trials.

**Pull request merged**